test/middleware/routing_middleware_test.dart

```diff
@@ -0,0 +1,423 @@
+import 'package:relic/relic.dart';
+import 'package:relic/src/adapter/context.dart';
+import 'package:relic/src/method/request_method.dart' as relic_req_method;
+import 'package:relic/src/middleware/routing_middleware.dart';
+import 'package:test/test.dart';
+
+// Simple fake implementations for testing
+class FakeRequest implements Request {
+  @override
+  final Uri url;
+  @override
+  final relic_req_method.RequestMethod method;
+
+  @override
+  dynamic noSuchMethod(final Invocation invocation) {
+    throw UnimplementedError(invocation.memberName.toString());
+  }
+
+  FakeRequest(final String path,
+      {this.method = relic_req_method.RequestMethod.get})
+      : url = Uri.parse('http://localhost$path');
+}
+
+class FakeResponse implements Response {
+  @override
+  final int statusCode;
+  final Map<Symbol, String>? parameters;
+
+  FakeResponse(this.statusCode, {this.parameters});
+
+  @override
+  dynamic noSuchMethod(final Invocation invocation) {
+    throw UnimplementedError(invocation.memberName.toString());
+  }
+}
+
+void main() {
+  group('RoutingMiddleware', () {
+    late Router<Handler> router;
+    late RoutingMiddleware middleware;
+
+    setUp(() {
+      router = Router<Handler>();
+      middleware = RoutingMiddleware(router);
+    });
+
+    group('Parameter Propagation', () {
+      test(
+          'Given a router with a parameterized route and RoutingMiddleware, '
+          'When a request matches the parameterized route, '
+          'Then the handler receives correct path parameters', () async {
+        Map<Symbol, String>? capturedParams;
+        Future<ResponseContext> testHandler(final RequestContext ctx) async {
+          capturedParams = ctx.pathParameters;
+          return (ctx as RespondableContext)
+              .withResponse(FakeResponse(200, parameters: capturedParams));
+        }
+
+        router.add(Method.get, '/users/:id', testHandler);
+
+        final initialCtx = FakeRequest('/users/123').toContext(Object());
+        final resultingCtx =
+            await middleware.meddle((final RequestContext ctx) async {
+          return (ctx as RespondableContext).withResponse(FakeResponse(404));
+        })(initialCtx);
+
+        expect(capturedParams, isNotNull);
+        expect(capturedParams, equals({#id: '123'}));
+        expect(resultingCtx, isA<ResponseContext>());
+        final response =
+            (resultingCtx as ResponseContext).response as FakeResponse;
+        expect(response.statusCode, equals(200));
+        expect(response.parameters, equals({#id: '123'}));
```

Comment:
Why thread the parameters through the response as well?

The `as FakeResponse` seems a bit weird, that the implementation object type has to be preserved even if it's not part of the interface.

I think asserting just on `capturedParams` is totally sufficient (and how I would write a similar test to see what the server was given).

Reply:
I no longer use the `_FakeResponse.parameters` in the test, and hence I have removed the class completely and just use the regular `Response` class.

Reply:
Nice, yeah, that simplifies comprehension in my eyes :)